design/005-Wire-Encoding-Abstraction.md

```diff
@@ -0,0 +1,118 @@
+# Proposal: Abstraction of Wire Messages Encoding
+
+* Author(s): Sebastian Stammler (@sebastianst)
+* Affected repository: go-perun
+
+## Summary
+
+The goal is to be able to choose different de/-serialization methods for wire
+messages when using go-perun's `wire` package for inter-node wire message
+exchange. After the changes, it should be possible to choose protobuf
+serialization.
+
+## Motivation
+
+Wire messages in go-perun currently use a home-baked and non-exchangeable
+encoding and decoding format. When developing new Perun channel client
+implementations, it would currently be necessary to mirror this custom
+serialization format.
+
+## Details
+
+Wire messages in go-perun currently use a home-baked and non-exchangeable
+encoding and decoding format, as specified by `Encode` and `Decode`
+methods on wire messages (see, e.g., message definitions in `client/*msgs.go`),
+methods `Envelope.(En|De)code` and functions `(En|De)code` in `wire/msg.go`.
+
+It should be possible to choose a different wire serialization protocol, e.g., protobuf.
+
+### Note on `pkg/io` serialization
+
+Note, thought, that this custom format is _not_ Go-specific. It is a packed,
+little-endian binary encoding format and the serialization logic of actual Go
+types is defined in package `pkg/io` and in particular in file
+`pkg/io/serialize.go`. (However, I just realize that `*big.Int`s are actually
+encoded in big-endian byte order...so we might as well switch to big-endianess
+in `pkg/io/serialize.go` to be more consistent
+([related issue](https://github.com/hyperledger-labs/go-perun/issues/167)).)
+
+## Rationale
+
+See #motivation.
+
+## Impact
+
+This can be implemented in a backwards-compatible fashion, where `pkg/io`
+encoding would be the default and other serialization backends could be injected
+at runtime.
+
+## Implementation
+
+Luckily, this can be achieved with a few changes to package `wire` and the
+`wire.Msg` implementations itself.
+
+The core idea is to create an envelope serialization backend, which can be set
+at runtime (e.g., using `init()` of the respecive serialization backend
+implementation). This backend's functions will then be used by
+`Envelope.(En|De)code` of package `wire`. The `pkg/io` serialization would then
+just be the default backend. Furthermore, the `Encode` methods of all `Msg`
+implementations would also be moved to this backend.
+
+Some of the proposed steps:
+* Create _Wire Serialization Backend_
+```go
+package wire
+
+// EnvelopeSerializer is a serialization backend for wire messages. It should be
+// set by the implementation packet's init() function.
```

Comment:
Should it not be `implementation package's init() fucntion` ? It currently says `packet's init() funtion`.